### ISSUE TITLE
unknown: GC crash on Windows with go-sqlite3

### ISSUE THREAD

**Issue by ToadKing** [golang/go#16294]
I've been getting weird fatal errors with Go 1.6.2 when doing queries on a SQLite database. They appear to be nearly completely random and we have not been able to get a case that reproduces it reliably yet, so I'm not sure how help this can provide.

They've been happening when using the https://github.com/mattn/go-sqlite3 package.

Relevant part of the stacktrace - the rest of it is our my code.

```
fatal error: unexpected signal during runtime execution
[signal 0xc0000005 code=0x0 addr=0xffffffffffffffff pc=0x40e8fd]

goroutine 381 [running]:
runtime.throw(0xd1e0c0, 0x2a)
	C:/Go/src/runtime/panic.go:547 +0x97 fp=0xc0839d3d58 sp=0xc0839d3d40
runtime.sigpanic()
	C:/Go/src/runtime/signal_windows.go:155 +0x68 fp=0xc0839d3d80 sp=0xc0839d3d58
runtime.getitab(0xafc840, 0x4057e8, 0xc082011001, 0x0)
	C:/Go/src/runtime/iface.go:83 +0x2ed fp=0xc0839d3e68 sp=0xc0839d3d80
runtime.assertE2I2(0xafc840, 0x4057e8, 0xc0820110e0, 0xc0839d3f30, 0x10)
	C:/Go/src/runtime/iface.go:364 +0x74 fp=0xc0839d3e90 sp=0xc0839d3e68
database/sql/driver.defaultConverter.ConvertValue(0x4057e8, 0xc0820110e0, 0x0, 0x0, 0x0, 0x0)
	C:/Go/src/database/sql/driver/types.go:223 +0xe8 fp=0xc0839d3fe0 sp=0xc0839d3e90
database/sql.driverArgs(0xc0839d42b0, 0xc0832205a8, 0x1, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
	C:/Go/src/database/sql/convert.go:36 +0x1e8 fp=0xc0839d41a8 sp=0xc0839d3fe0
database/sql.rowsiFromStatement(0x134a340, 0xc0821f93b0, 0x134a2d8, 0xc082948f60, 0xc0832205a8, 0x1, 0x1, 0x0, 0x0, 0x0, ...)
	C:/Go/src/database/sql/sql.go:1648 +0x2db fp=0xc0839d42b0 sp=0xc0839d41a8
database/sql.(*Stmt).Query(0xc082508300, 0xc0832205a8, 0x1, 0x1, 0x0, 0x0, 0x0)
	C:/Go/src/database/sql/sql.go:1611 +0x2bd fp=0xc0839d43d0 sp=0xc0839d42b0
```

go version/env

```
go version go1.6 windows/amd64

set GOARCH=amd64
set GOBIN=
set GOEXE=.exe
set GOHOSTARCH=amd64
set GOHOSTOS=windows
set GOOS=windows
set GOPATH=
set GORACE=
set GOROOT=C:\Go
set GOTOOLDIR=C:\Go\pkg\tool\windows_amd64
set GO15VENDOREXPERIMENT=1
set CC=gcc
set GOGCCFLAGS=-m64 -mthreads -fmessage-length=0
set CXX=g++
set CGO_ENABLED=1
```

**Comment by mattn** [comment]
what the code did you run?

**Comment by ToadKing** [comment]
I can't post the whole thing since it's proprietary but it's basically this:

```go
conn, err := sql.Open("sqlite3", "db.sqlite")
if err != nil {
	return err
}

stmt, err := conn.Prepare(query)
if err != nil {
	return err
}
defer stmt.Close()

rows, _ := stmt.Query(input...)
defer rows.Close()
for rows.Next() {
	r, err := transformFunc(rows)

	if err != nil {
		return err
	}

	results = append(results, r)
}
return results
```

It crashes in `stmt.Query(input...)`. `input` is `[]interface{}`.

**Comment by ianlancetaylor** [comment]
This looks like memory corruption.  Have you tried running your program under the race detector?  See https://blog.golang.org/race-detector .

**Comment by ToadKing** [comment]
I was able to detect some race conditions in the program and fix them but the crash still happens rarely. For some reason the crash only happens in the database code though.

```
goroutine 527 [running]:
runtime.throw(0xc52bc0, 0x5)
	c:/go/src/runtime/panic.go:547 +0x97 fp=0xc083e7be20 sp=0xc083e7be08
runtime.sigpanic()
	c:/go/src/runtime/signal_windows.go:164 +0x12a fp=0xc083e7be48 sp=0xc083e7be20
runtime.efacethash(0x3a2272227b3a2274, 0x2c303a2267222c30, 0x2c302c302c302c30)
	c:/go/src/runtime/iface.go:414 +0x15 fp=0xc083e7be50 sp=0xc083e7be48
database/sql/driver.IsScanValue(0x3a2272227b3a2274, 0x2c303a2267222c30, 0x10)
	c:/go/src/database/sql/driver/types.go:192 +0x53 fp=0xc083e7be80 sp=0xc083e7be50
database/sql/driver.IsValue(0x3a2272227b3a2274, 0x2c303a2267222c30, 0x2c302c302c302c30)
	c:/go/src/database/sql/driver/types.go:177 +0x32 fp=0xc083e7bea0 sp=0xc083e7be80
database/sql/driver.defaultConverter.ConvertValue(0x3a2272227b3a2274, 0x2c303a2267222c30, 0x0, 0x0, 0x0, 0x0)
	c:/go/src/database/sql/driver/types.go:219 +0x6b fp=0xc083e7bff0 sp=0xc083e7bea0
database/sql.driverArgs(0xc083e7c2c0, 0xc0838365b8, 0x1, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
	c:/go/src/database/sql/convert.go:36 +0x1e8 fp=0xc083e7c1b8 sp=0xc083e7bff0
database/sql.rowsiFromStatement(0x15a0230, 0xc082aa0af0, 0x15a01c8, 0xc0827864b0, 0xc0838365b8, 0x1, 0x1, 0x0, 0x0, 0x0, ...)
	c:/go/src/database/sql/sql.go:1648 +0x2db fp=0xc083e7c2c0 sp=0xc083e7c1b8
database/sql.(*Stmt).Query(0xc0834d6100, 0xc0838365b8, 0x1, 0x1, 0x0, 0x0, 0x0)
	c:/go/src/database/sql/sql.go:1611 +0x2bd fp=0xc083e7c3e0 sp=0xc083e7c2c0
```

**Comment by ToadKing** [comment]
We are still having issues in Go 1.7.1. Different crash type but still in the database/sql package.

```
runtime: pointer 0xc048570d18 to unallocated spanidx=0x32b8 span.base()=0xc04856a000 span.limit=0xc04b837f80 span.state=3
runtime: found in object at *(0xc0432deb40+0x18)
object=0xc0432deb40 k=0x602196f s.base()=0xc0432de000 s.limit=0xc0432e0000 s.sizeclass=5 s.elemsize=64
 *(object+0) = 0x644f90
 *(object+8) = 0xacae7a
 *(object+16) = 0x5a
 *(object+24) = 0xc048570d18 <==
 *(object+32) = 0x2
 *(object+40) = 0x2
 *(object+48) = 0xb07f50
 *(object+56) = 0xc0424e2260
fatal error: found bad pointer in Go heap (incorrect use of unsafe or cgo?)

goroutine 31 [running]:
runtime.throw(0xac57a6, 0x3e)
	C:/Go/src/runtime/panic.go:566 +0x9c fp=0xc042449e18 sp=0xc042449df8
runtime.heapBitsForObject(0xc048570d18, 0xc0432deb40, 0x18, 0xc041fc16c0, 0xc000000000, 0x3bb57d8, 0xc042017228, 0x41c956)
	C:/Go/src/runtime/mbitmap.go:429 +0x2c1 fp=0xc042449e70 sp=0xc042449e18
runtime.scanobject(0xc0432deb40, 0xc042017228)
	C:/Go/src/runtime/mgcmark.go:1169 +0x177 fp=0xc042449ef0 sp=0xc042449e70
runtime.gcDrain(0xc042017228, 0x6)
	C:/Go/src/runtime/mgcmark.go:993 +0x16a fp=0xc042449f28 sp=0xc042449ef0
runtime.gcBgMarkWorker(0xc042016000)
	C:/Go/src/runtime/mgc.go:1450 +0x4d7 fp=0xc042449f98 sp=0xc042449f28
runtime.goexit()
	C:/Go/src/runtime/asm_amd64.s:2086 +0x1 fp=0xc042449fa0 sp=0xc042449f98
created by runtime.gcBgMarkStartWorkers
	C:/Go/src/runtime/mgc.go:1341 +0x9f

...

goroutine 3690 [runnable]:
database/sql.convertAssign(0x9c0e60, 0xc04391ef18, 0x9cb500, 0xc049fdf300, 0x0, 0x0)
	C:/Go/src/database/sql/convert.go:110 +0x1f94
database/sql.(*Rows).Scan(0xc0445e49c0, 0xc04857ac28, 0xc, 0xc, 0xc04857acf0, 0xc04857ad88)
	C:/Go/src/database/sql/sql.go:1850 +0xc9

...
```

The race condition checker isn't turning up anything.

**Comment by bradfitz** [comment]
The error has nothing to do with `database/sql`. The error you're getting is:

`fatal error: found bad pointer in Go heap (incorrect use of unsafe or cgo?)`

Which begs the question: do you have any incorrect use of unsafe or cgo? database/sql doesn't use either.

**Comment by ToadKing** [comment]
I've narrowed down the issue more, and as such I can start posting examples.

The bug seems to appear in our database code that queries rows and transforms the results to structures. We have another trace that has the object with the bad pointer in it:

```
runtime: pointer 0xc0442bf570 to unallocated spanidx=0x115f span.base()=0xc0442be000 span.limit=0xc04431c000 span.state=2
runtime: found in object at *(0xc0426ee410+0x18)
object=0xc0426ee410 k=0x6021377 s.base()=0xc0426ee000 s.limit=0xc0426effe0 s.sizeclass=6 s.elemsize=80
 *(object+0) = 0x642350
 *(object+8) = 0xac469f
 *(object+16) = 0x44
 *(object+24) = 0xc0442bf570 <==
 *(object+32) = 0x2
 *(object+40) = 0x2
 *(object+48) = 0xc0442bf560
 *(object+56) = 0x1
 *(object+64) = 0x1
 *(object+72) = 0x0
fatal error: found bad pointer in Go heap (incorrect use of unsafe or cgo?)

...

goroutine 55 [GC assist wait]:
github.com/mattn/go-sqlite3.(*SQLiteConn).Prepare(0xc042070eb0, 0xd22235, 0x5a, 0x0, 0x0, 0x0, 0x0)
        D:/Repos/SSEnext-Core/src/github.com/mattn/go-sqlite3/sqlite3.go:731 +0x11e
database/sql.(*driverConn).prepareLocked(0xc04220f490, 0xd22235, 0x5a, 0x0, 0x0, 0x457c60, 0xc042278000)
        D:/Go/src/database/sql/sql.go:301 +0x90
database/sql.(*DB).prepare(0xc0420f2630, 0xd22235, 0x5a, 0x1, 0xc042450000, 0xc0427a4820, 0x7)
        D:/Go/src/database/sql/sql.go:987 +0xca
database/sql.(*DB).Prepare(0xc0420f2630, 0xd22235, 0x5a, 0xc042450017, 0x3, 0x40)
        D:/Go/src/database/sql/sql.go:964 +0x7d
xxx/db.QueryRows.func1(0xc042066140, 0x0, 0x0)
        D:/Repos/SSEnext-Core/src/github.com/steelseries/ssenext/db/db.go:134 +0xe4
xxx/db.ExecuteDBFunc(0xc04285e180, 0x0, 0x0)
        D:/Repos/SSEnext-Core/src/github.com/steelseries/ssenext/db/db.go:23 +0xc5
xxx/db.QueryRows(0xd22235, 0x5a, 0xc043b14bf8, 0x2, 0x2, 0xd96190, 0x0, 0x0, 0xc042b558e8, 0xc042b55918, ...)
```

The `0xc0426ee410` object is actually a function. `ExecuteDBFunc` and `QueryRows` look like this:

```
type DBFunc func(*connector.Connection) error
type RowTransformFunc func(*sql.Rows) (interface{}, error)

func ExecuteDBFunc(f DBFunc) error {
	runtime.GC()
	conn := connector.GetDBConnection()
	defer conn.Release()
	return f(conn)
}

func QueryRows(query string, input []interface{}, transformFunc RowTransformFunc) (results []interface{}, err error) {

	f := func(conn *connector.Connection) error {
		stmt, err := conn.Prepare(query)
		if err != nil {
			return err
		}
		defer stmt.Close()

		rows, _ := stmt.Query(input...)
		defer rows.Close()
		for rows.Next() {
			r, err := transformFunc(rows)

			if err != nil {
				return err
			}

			results = append(results, r)
		}
		return nil
	}

	err = ExecuteDBFunc(f)
	return
}
``` 

connection.Connection is just a wrapper around a `sql.DB` object. (No idea why it's wrapped, probably some legacy thing):

```
type Connection struct {
	*sql.DB
}
```

Does any of this look suspicious?

**Comment by mwhudson** [comment]
Looks a bit similar to #17318?

**Comment by ToadKing** [comment]
Yes, it does look similar. Especially with the comment that it didn't happen in Go 1.5, but does in 1.6 and 1.7 which is the same issue I'm getting. I can work around it locally for the time being.

I'll close this issue for now, and if it still happens after #17318 is fixed I'll reopen or create a new issue.
